File: src/worker.ts
#!/usr/bin/env node
/**
 * clrun worker — detached background process that manages a single PTY session.
 *
 * Spawned by `clrun run`. Runs until the PTY exits, is killed, or is suspended.
 *
 * Args: <terminalId> <command> <cwd> <projectRoot> [--restore]
 */

import * as os from 'os';
import * as fs from 'fs';
import * as path from 'path';
import { appendToBuffer, initBuffer } from './buffer/buffer-manager';
import { getNextQueued, markSent, pendingCount } from './queue/queue-engine';
import { writeSession, readSession, updateSession, detectShell } from './pty/pty-manager';
import { logEvent } from './ledger/ledger';
import { getClrunPaths } from './utils/paths';
import type { SessionMetadata, SavedState } from './types';

// ─── Configuration ──────────────────────────────────────────────────────────

const IDLE_TIMEOUT_MS = 5 * 60 * 1000;  // 5 minutes
const IDLE_CHECK_INTERVAL_MS = 30 * 1000; // check every 30s
const SUSPEND_CAPTURE_WAIT_MS = 600;      // wait for shell to flush state files

// ─── Parse Arguments ────────────────────────────────────────────────────────

const args = process.argv.slice(2);
const restoreFlag = args.includes('--restore');
const positional = args.filter((a) => a !== '--restore');
const [terminalId, command, cwd, projectRoot] = positional;

if (!terminalId || !command || !cwd || !projectRoot) {
  process.stderr.write('worker: missing arguments\n');
  process.exit(1);
}

// ─── Import node-pty dynamically ────────────────────────────────────────────

let pty: typeof import('node-pty');
try {
  pty = require('node-pty');
} catch {
  process.stderr.write('worker: node-pty not available. Install with: npm install node-pty\n');
  process.exit(1);
}

// ─── Internal vars to skip during env restore ───────────────────────────────

const SKIP_ENV_VARS = new Set([
  '_', 'SHLVL', 'PWD', 'OLDPWD', 'SHELL', 'TERM', 'TERM_PROGRAM',
  'TERM_PROGRAM_VERSION', 'TERM_SESSION_ID', 'TMPDIR', 'LOGNAME', 'USER',
  'HOME', 'LANG', 'SSH_AUTH_SOCK',
]);

// ─── Resolve restore state ──────────────────────────────────────────────────

let restoreState: SavedState | null = null;
let restoreCwd = cwd;

if (restoreFlag) {
  const session = readSession(terminalId, projectRoot);
  if (session?.saved_state) {
    restoreState = session.saved_state;
    restoreCwd = restoreState.cwd;
  }
}

// ─── Spawn PTY ──────────────────────────────────────────────────────────────

const shell = detectShell();

const ptyProcess = pty.spawn(shell, [], {
  name: 'xterm-256color',
  cols: 120,
  rows: 40,
  cwd: restoreCwd,
  env: { ...process.env } as Record<string, string>,
});

const ptyPid = ptyProcess.pid;

// ─── Restore or send initial command ────────────────────────────────────────

setTimeout(() => {
  if (restoreState) {
    // Re-export saved environment variables
    const exports: string[] = [];
    for (const [key, value] of Object.entries(restoreState!.env)) {
      if (SKIP_ENV_VARS.has(key)) continue;
      // Escape single quotes in value
      const escaped = value.replace(/'/g, "'\\''");
      exports.push(`export ${key}='${escaped}'`);
    }
    if (exports.length > 0) {
      ptyProcess.write(exports.join(' && ') + '\n');
    }

    // Append a restore marker to the buffer
    appendToBuffer(terminalId, '\n--- session restored ---\n', projectRoot);

    logEvent('session.restored', projectRoot, terminalId, {
      restored_cwd: restoreState!.cwd,
      restored_vars: Object.keys(restoreState!.env).filter((k) => !SKIP_ENV_VARS.has(k)).length,
    });
  } else {
    // Normal start: send the initial command
    ptyProcess.write(command + '\n');
  }
}, 80);

// ─── Initialize State ───────────────────────────────────────────────────────

if (!restoreFlag) {
  initBuffer(terminalId, projectRoot);
}

const now = new Date().toISOString();
const session: SessionMetadata = {
  terminal_id: terminalId,
  created_at: restoreFlag
    ? (readSession(terminalId, projectRoot)?.created_at ?? now)
    : now,
  cwd: restoreCwd,
  command,
  shell,
  status: 'running',
  pid: ptyPid,
  worker_pid: process.pid,
  queue_length: 0,
  last_exit_code: null,
  last_activity_at: now,
};
writeSession(session, projectRoot);

if (!restoreFlag) {
  logEvent('session.created', projectRoot, terminalId, { command, cwd, pid: ptyPid });
}

// ─── Idle Tracking ──────────────────────────────────────────────────────────

let lastActivity = Date.now();

function resetIdle(): void {
  lastActivity = Date.now();
}

// ─── Buffer Output ──────────────────────────────────────────────────────────

ptyProcess.onData((data: string) => {
  appendToBuffer(terminalId, data, projectRoot);
  resetIdle();
  updateSession(
    terminalId,
    { last_activity_at: new Date().toISOString() },
    projectRoot
  );
});

// ─── Queue Processing ───────────────────────────────────────────────────────

let queueInterval: ReturnType<typeof setInterval>;

// Raw prefix marker: inputs starting with this are sent verbatim (no \r appended).
// Used by `clrun key` to send escape sequences for arrow keys, tab, etc.
const RAW_PREFIX = '\x00RAW\x00';

function processQueue(): void {
  try {
    let next = getNextQueued(terminalId, projectRoot);
    while (next) {
      if (next.input.startsWith(RAW_PREFIX)) {
        // Raw mode: send the byte sequence verbatim, no \r appended.
        const raw = next.input.slice(RAW_PREFIX.length);
        ptyProcess.write(raw);
      } else {
        // Normal mode: use \r (carriage return) instead of \n (line feed).
        // In cooked mode the terminal driver maps \r → \n (ICRNL) so this
        // works identically to \n for regular shells.  In raw mode (used by
        // TUI frameworks like @clack/prompts, inquirer, etc.) \r is the
        // correct "Enter" key — \n would be ignored or misinterpreted.
        ptyProcess.write(next.input + '\r');
      }
      markSent(terminalId, next.queue_id, projectRoot);
      resetIdle();
      logEvent('input.sent', projectRoot, terminalId, {
        queue_id: next.queue_id,
        input: next.input.startsWith(RAW_PREFIX) ? '[raw keys]' : next.input,
      });
      next = getNextQueued(terminalId, projectRoot);
    }

    const count = pendingCount(terminalId, projectRoot);
    updateSession(terminalId, { queue_length: count }, projectRoot);
  } catch {
    // Non-fatal — retry on next interval
  }
}

// Poll queue every 200ms
queueInterval = setInterval(processQueue, 200);

// Also respond to SIGUSR1 for immediate queue processing
process.on('SIGUSR1', () => {
  resetIdle();
  processQueue();
});

// ─── Idle Suspension ────────────────────────────────────────────────────────

let suspending = false;

const idleCheckInterval = setInterval(async () => {
  if (suspending) return;

  const idle = Date.now() - lastActivity;
  if (idle < IDLE_TIMEOUT_MS) return;

  suspending = true;
  clearInterval(queueInterval);
  clearInterval(idleCheckInterval);

  try {
    await captureAndSuspend();
  } catch {
    // If capture fails, just kill without saving state
    updateSession(terminalId, { status: 'suspended', last_activity_at: new Date().toISOString() }, projectRoot);
    logEvent('session.suspended', projectRoot, terminalId, { capture_failed: true });
    try { ptyProcess.kill(); } catch {}
    process.exit(0);
  }
}, IDLE_CHECK_INTERVAL_MS);

async function captureAndSuspend(): Promise<void> {
  const paths = getClrunPaths(projectRoot);
  const cwdFile = path.join(paths.sessionsDir, `${terminalId}.state.cwd`);
  const envFile = path.join(paths.sessionsDir, `${terminalId}.state.env`);

  // Clean up any stale state files
  try { fs.unlinkSync(cwdFile); } catch {}
  try { fs.unlinkSync(envFile); } catch {}

  // Ask the shell to dump its state
  ptyProcess.write(`pwd > '${cwdFile}'\n`);
  ptyProcess.write(`env -0 > '${envFile}'\n`);

  // Wait for the shell to flush
  await new Promise((r) => setTimeout(r, SUSPEND_CAPTURE_WAIT_MS));

  // Read captured state — fall back to the session's original cwd.
  let capturedCwd = cwd;
  const capturedEnv: Record<string, string> = {};

  try {
    capturedCwd = fs.readFileSync(cwdFile, 'utf-8').trim();
  } catch {
    // Fall back to original cwd
  }

  try {
    const rawEnv = fs.readFileSync(envFile, 'utf-8');
    // env -0 separates entries with null bytes
    const entries = rawEnv.split('\0').filter(Boolean);
    for (const entry of entries) {
      const eqIdx = entry.indexOf('=');
      if (eqIdx > 0) {
        const key = entry.slice(0, eqIdx);
        const value = entry.slice(eqIdx + 1);
        capturedEnv[key] = value;
      }
    }
  } catch {
    // If env capture fails, save empty env
  }

  // Clean up temp files
  try { fs.unlinkSync(cwdFile); } catch {}
  try { fs.unlinkSync(envFile); } catch {}

  const savedState: SavedState = {
    cwd: capturedCwd,
    env: capturedEnv,
    captured_at: new Date().toISOString(),
  };

  // Update session with suspended status and saved state
  updateSession(
    terminalId,
    {
      status: 'suspended',
      last_activity_at: new Date().toISOString(),
      saved_state: savedState,
    },
    projectRoot
  );

  logEvent('session.suspended', projectRoot, terminalId, {
    saved_cwd: capturedCwd,
    saved_env_count: Object.keys(capturedEnv).length,
  });

  appendToBuffer(terminalId, '\n--- session suspended (idle timeout) ---\n', projectRoot);

  // Kill the PTY
  try { ptyProcess.kill(); } catch {}

  setTimeout(() => process.exit(0), 100);
}

// ─── PTY Exit ───────────────────────────────────────────────────────────────

ptyProcess.onExit(({ exitCode }: { exitCode: number }) => {
  if (suspending) return; // Handled by captureAndSuspend

  clearInterval(queueInterval);
  clearInterval(idleCheckInterval);

  updateSession(
    terminalId,
    {
      status: 'exited',
      last_exit_code: exitCode,
      last_activity_at: new Date().toISOString(),
      queue_length: 0,
    },
    projectRoot
  );

  logEvent('session.exited', projectRoot, terminalId, { exit_code: exitCode });

  setTimeout(() => {
    process.exit(0);
  }, 100);
});

// ─── Graceful Shutdown ──────────────────────────────────────────────────────

function shutdown(signal: string): void {
  clearInterval(queueInterval);
  clearInterval(idleCheckInterval);

  try {
    ptyProcess.kill();
  } catch {
    // Already dead
  }

  updateSession(
    terminalId,
    {
      status: 'killed',
      last_activity_at: new Date().toISOString(),
    },
    projectRoot
  );

  logEvent('session.killed', projectRoot, terminalId, { signal });
  process.exit(0);
}

process.on('SIGTERM', () => shutdown('SIGTERM'));
process.on('SIGINT', () => shutdown('SIGINT'));

File: python/clrun/worker.py
#!/usr/bin/env python3
"""
clrun worker — detached background process that manages a single PTY session.

Spawned by `clrun run`. Runs until the PTY exits, is killed, or is suspended.

Usage: python -m clrun.worker <terminalId> <command> <cwd> <projectRoot> [--restore]
"""

from __future__ import annotations

import os
import select
import signal
import sys
import time
from datetime import datetime, timezone

import pexpect

from clrun.buffer.buffer_manager import append_to_buffer, init_buffer
from clrun.queue.queue_engine import get_next_queued, mark_sent, pending_count
from clrun.pty.pty_manager import write_session, read_session, update_session, detect_shell
from clrun.ledger.ledger import log_event
from clrun.utils.paths import get_clrun_paths, ensure_clrun_dirs
from clrun.types import SessionMetadata, SavedState

# ─── Configuration ───────────────────────────────────────────────────────────

IDLE_TIMEOUT_S = 5 * 60       # 5 minutes
SUSPEND_CAPTURE_WAIT_S = 0.6  # wait for shell to flush state files
RAW_PREFIX = "\x00RAW\x00"

SKIP_ENV_VARS = {
    "_", "SHLVL", "PWD", "OLDPWD", "SHELL", "TERM", "TERM_PROGRAM",
    "TERM_PROGRAM_VERSION", "TERM_SESSION_ID", "TMPDIR", "LOGNAME", "USER",
    "HOME", "LANG", "SSH_AUTH_SOCK",
}

# ─── State ───────────────────────────────────────────────────────────────────

last_activity = time.time()
suspending = False
child: pexpect.spawn | None = None
sigusr1_received = False


def reset_idle() -> None:
    global last_activity
    last_activity = time.time()


def now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


def main() -> None:
    global child, suspending, sigusr1_received

    args = sys.argv[1:]
    restore_flag = "--restore" in args
    positional = [a for a in args if a != "--restore"]

    if len(positional) < 4:
        sys.stderr.write("worker: missing arguments\n")
        sys.exit(1)

    terminal_id, command, cwd, project_root = positional[:4]

    # ─── Ensure directories exist ────────────────────────────────────────
    ensure_clrun_dirs(project_root)

    # ─── Resolve restore state ───────────────────────────────────────────
    restore_state: SavedState | None = None
    restore_cwd = cwd

    if restore_flag:
        session = read_session(terminal_id, project_root)
        if session and session.saved_state:
            restore_state = session.saved_state
            restore_cwd = restore_state.cwd

    # ─── Spawn PTY ───────────────────────────────────────────────────────
    shell = detect_shell()
    env = dict(os.environ)
    env["TERM"] = "xterm-256color"

    child = pexpect.spawn(
        shell,
        encoding="utf-8",
        cwd=restore_cwd,
        env=env,
        dimensions=(40, 120),
        timeout=None,
    )
    pty_pid = child.pid

    # ─── Initialize state ────────────────────────────────────────────────
    if not restore_flag:
        init_buffer(terminal_id, project_root)

    existing = read_session(terminal_id, project_root) if restore_flag else None
    session_data = SessionMetadata(
        terminal_id=terminal_id,
        created_at=existing.created_at if existing else now_iso(),
        cwd=restore_cwd,
        command=command,
        shell=shell,
        status="running",
        pid=pty_pid,
        worker_pid=os.getpid(),
        queue_length=0,
        last_exit_code=None,
        last_activity_at=now_iso(),
    )
    write_session(session_data, project_root)

    if not restore_flag:
        log_event("session.created", project_root, terminal_id, {"command": command, "cwd": cwd, "pid": pty_pid})

    # ─── Send initial command or restore ─────────────────────────────────
    time.sleep(0.08)

    if restore_state:
        exports = []
        for key, value in restore_state.env.items():
            if key in SKIP_ENV_VARS:
                continue
            escaped = value.replace("'", "'\\''")
            exports.append(f"export {key}='{escaped}'")
        if exports:
            child.sendline(" && ".join(exports))
        append_to_buffer(terminal_id, "\n--- session restored ---\n", project_root)
        restored_vars = len([k for k in restore_state.env if k not in SKIP_ENV_VARS])
        log_event("session.restored", project_root, terminal_id, {
            "restored_cwd": restore_state.cwd,
            "restored_vars": restored_vars,
        })
    else:
        child.sendline(command)

    # ─── SIGUSR1 handler for immediate queue processing ──────────────────
    def sigusr1_handler(signum: int, frame: object) -> None:
        global sigusr1_received
        sigusr1_received = True

    signal.signal(signal.SIGUSR1, sigusr1_handler)

    # ─── Helper: read available PTY output ───────────────────────────────
    def drain_output() -> None:
        """Read all available data from the PTY and append to buffer."""
        try:
            fd = child.fileno()
        except Exception:
            return
        while True:
            try:
                rlist, _, _ = select.select([fd], [], [], 0)
                if not rlist:
                    break
                data = child.read_nonblocking(size=4096, timeout=0)
                if data:
                    append_to_buffer(terminal_id, data, project_root)
                    reset_idle()
                else:
                    break
            except pexpect.TIMEOUT:
                break
            except pexpect.EOF:
                break
            except Exception:
                break

    # ─── Helper: process queue ───────────────────────────────────────────
    def process_queue() -> None:
        try:
            entry = get_next_queued(terminal_id, project_root)
            while entry:
                if entry.input.startswith(RAW_PREFIX):
                    raw = entry.input[len(RAW_PREFIX):]
                    child.send(raw)
                else:
                    child.sendline(entry.input)
                mark_sent(terminal_id, entry.queue_id, project_root)
                reset_idle()
                log_event("input.sent", project_root, terminal_id, {
                    "queue_id": entry.queue_id,
                    "input": "[raw keys]" if entry.input.startswith(RAW_PREFIX) else entry.input,
                })
                entry = get_next_queued(terminal_id, project_root)
            count = pending_count(terminal_id, project_root)
            update_session(terminal_id, {"queue_length": count}, project_root)
        except Exception:
            pass

    # ─── Capture and suspend ─────────────────────────────────────────────
    def capture_and_suspend() -> None:
        global suspending
        paths = get_clrun_paths(project_root)
        cwd_file = os.path.join(paths.sessions_dir, f"{terminal_id}.state.cwd")
        env_file = os.path.join(paths.sessions_dir, f"{terminal_id}.state.env")

        for f in [cwd_file, env_file]:
            try:
                os.unlink(f)
            except OSError:
                pass

        child.sendline(f"pwd > '{cwd_file}'")
        child.sendline(f"env -0 > '{env_file}'")
        time.sleep(SUSPEND_CAPTURE_WAIT_S)

        captured_cwd = cwd
        captured_env: dict[str, str] = {}

        try:
            with open(cwd_file, "r", encoding="utf-8") as f:
                captured_cwd = f.read().strip()
        except Exception:
            pass

        try:
            with open(env_file, "r", encoding="utf-8") as f:
                raw = f.read()
            for entry in raw.split("\0"):
                if not entry:
                    continue
                eq = entry.find("=")
                if eq > 0:
                    captured_env[entry[:eq]] = entry[eq + 1:]
        except Exception:
            pass

        for f in [cwd_file, env_file]:
            try:
                os.unlink(f)
            except OSError:
                pass

        saved_state = {
            "cwd": captured_cwd,
            "env": captured_env,
            "captured_at": now_iso(),
        }

        update_session(terminal_id, {
            "status": "suspended",
            "last_activity_at": now_iso(),
            "saved_state": saved_state,
        }, project_root)

        log_event("session.suspended", project_root, terminal_id, {
            "saved_cwd": captured_cwd,
            "saved_env_count": len(captured_env),
        })

        append_to_buffer(terminal_id, "\n--- session suspended (idle timeout) ---\n", project_root)

        try:
            child.terminate(force=True)
        except Exception:
            pass

    # ─── Graceful shutdown ───────────────────────────────────────────────
    def shutdown(signum: int = 0, frame: object = None) -> None:
        try:
            child.terminate(force=True)
        except Exception:
            pass
        update_session(terminal_id, {"status": "killed", "last_activity_at": now_iso()}, project_root)
        log_event("session.killed", project_root, terminal_id, {"signal": signum})
        sys.exit(0)

    signal.signal(signal.SIGTERM, shutdown)
    signal.signal(signal.SIGINT, shutdown)

    # ─── Main event loop (single-threaded) ───────────────────────────────
    last_session_update = time.time()

    try:
        while True:
            # 1. Drain all available PTY output
            drain_output()

            # 2. Check if child is still alive
            if not child.isalive():
                # Drain any final output
                drain_output()
                break

            # 3. Process queue (on signal or every iteration)
            if sigusr1_received:
                sigusr1_received = False
                reset_idle()
            process_queue()

            # 4. Update session activity periodically (every 5s)
            now = time.time()
            if now - last_session_update > 5:
                update_session(terminal_id, {"last_activity_at": now_iso()}, project_root)
                last_session_update = now

            # 5. Check idle timeout
            idle = time.time() - last_activity
            if idle >= IDLE_TIMEOUT_S and not suspending:
                suspending = True
                try:
                    capture_and_suspend()
                except Exception:
                    update_session(terminal_id, {"status": "suspended", "last_activity_at": now_iso()}, project_root)
                    log_event("session.suspended", project_root, terminal_id, {"capture_failed": True})
                    try:
                        child.terminate(force=True)
                    except Exception:
                        pass
                sys.exit(0)

            # 6. Sleep briefly to avoid busy-waiting
            time.sleep(0.1)

    except Exception:
        pass

    # ─── PTY exited normally ─────────────────────────────────────────────
    if not suspending:
        exit_code = child.exitstatus if child.exitstatus is not None else child.signalstatus or 0
        update_session(terminal_id, {
            "status": "exited",
            "last_exit_code": exit_code,
            "last_activity_at": now_iso(),
            "queue_length": 0,
        }, project_root)
        log_event("session.exited", project_root, terminal_id, {"exit_code": exit_code})

    sys.exit(0)


if __name__ == "__main__":
    main()
